compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5079,6 +5066,47 @@ public void visit(BLangBinaryExpr binaryExpr, AnalyzerData data) {
         data.resultType = types.checkType(binaryExpr, actualType, data.expType);
     }
 
+    private BType getBinaryOperandType(BLangExpression operand, BLangBinaryExpr binaryExpr, BType referredExpType,
+                                       SymbolEnv env, AnalyzerData data) {
+        if ((referredExpType.tag == TypeTags.FLOAT || referredExpType.tag == TypeTags.DECIMAL ||
+                isOptionalFloatOrDecimal(referredExpType)) && isNumericLiteral(operand)) {
+            return checkAndGetType(operand, env, binaryExpr, data);
+        }
+        if (isBinaryExpr(operand)) {
```

Comment:
Basically, if the expr is not a `numeric-literal` we have to type check using `no-type` as expected type. Then only we can get the error `operator not defined ...` other than `incompatible types ...`. That is the purpose of the last if block.
Let's say we have an expression like `(10.0 + 5) - 2`. In the first iteration, type checking happens for `10.0 + 5` with no type, and in 2nd iteration, it will not go to the first if block as expected because there is not an expected type. 
In order to solve this problem, I added the mentioned if block.